### PR TITLE
Meta device initialization for FSDP in Trainer

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Same approach as in #18122

```py
trainer = Trainer(strategy="fsdp")

with trainer.init_module(empty_init=True):
    model = Model() # the model is now on the meta device (no memory occupied)
...
# materialization and param init and sharding happens in `strategy.setup`
trainer.fit(model)
```

This allows you to instantiate very large models that wouldn't fit in memory (either CPU or GPU) as fast as possible. No memory for weights get allocated, neither in CPU nor GPU memory and parameters are materialized/initialized with random weights directly at the time the model gets wrapped and sharded in `FSDPStrategy.setup()`.

**Notes:**
- This new feature is possible thanks to https://github.com/pytorch/pytorch/issues/104187 in PyTorch 2.1 nightly.
Requirement: Your submodules define a `reset_parameters()` method that can be called to init the params. This is the case for all (most) built-in PyTorch layers. If you have a custom layer, you'd have to add that method. This PR also updates our Trainsformers example that needs this `reset_parameters()` fixed.
- Documentation will be updated in a follow-up, both for Fabric and Trainer.

cc @borda @awaelchli @carmocca

### CODE REVIEW

File: src/lightning/pytorch/demos/transformer.py [hunk 1]
```diff
@@ -14,8 +14,12 @@
 import torch.nn as nn
 import torch.nn.functional as F
 from torch import Tensor
+from torch.nn.modules import MultiheadAttention
 from torch.utils.data import Dataset
 
+if hasattr(MultiheadAttention, "_reset_parameters") and not hasattr(MultiheadAttention, "reset_parameters"):
+    MultiheadAttention.reset_parameters = MultiheadAttention._reset_parameters
```
Comment (awaelchli) [inline src/lightning/pytorch/demos/transformer.py:21]: TODO: open issue in torch

File: src/lightning/pytorch/demos/transformer.py [hunk 2]
```diff
@@ -60,20 +64,29 @@ class PositionalEncoding(nn.Module):
...
-        div_term = torch.exp(torch.arange(0, dim, 2).float() * (-math.log(10000.0) / dim))
-        pe[:, 0::2] = torch.sin(position * div_term)
-        pe[:, 1::2] = torch.cos(position * div_term)
-        pe = pe.unsqueeze(0).transpose(0, 1)
+        pe = self._init_pos_encoding()
         # workaround, can't use buffer, see https://github.com/pytorch/pytorch/issues/68407
         self.register_parameter("pe", nn.Parameter(pe, requires_grad=False))
 
+    def reset_parameters(self) -> None:
+        self.pe.copy_(self._init_pos_encoding())
```
Comment (awaelchli) [inline src/lightning/pytorch/demos/transformer.py:75]: mypy has issues with this line of code 🤔

File: src/lightning/pytorch/demos/transformer.py (hunk 1 above)
Comment (awaelchli) [reply src/lightning/pytorch/demos/transformer.py:21]: https://github.com/pytorch/pytorch/issues/107909

File: src/lightning/pytorch/demos/transformer.py (hunk 2 above)
Comment (Borda) [reply src/lightning/pytorch/demos/transformer.py:76]: lets add ignore :)
```suggestion
        self.pe.copy_(self._init_pos_encoding())  # type: ignore
```

**Review by Borda — APPROVED** [review]

File: src/lightning/pytorch/strategies/fsdp.py [hunk 3]
```diff
@@ -267,7 +269,11 @@ def _setup_model(self, model: Module) -> Module:
         from torch.distributed.fsdp import FullyShardedDataParallel
 
         if any(isinstance(mod, FullyShardedDataParallel) for mod in model.modules()):
-            # the user wrapped at least one layer in `configure_model` already
+            # The user has wrapped their submodules manually, don't apply the auto wrap policy.
```
Comment (carmocca) [inline src/lightning/pytorch/strategies/fsdp.py:272]: This comment should go with the if below

**Review by carmocca — APPROVED** [review]

File: src/lightning/pytorch/strategies/fsdp.py (hunk 3 above)
Comment (awaelchli) [reply src/lightning/pytorch/strategies/fsdp.py:272]:
```suggestion
```

File: src/lightning/pytorch/strategies/fsdp.py [hunk 4]
```diff
@@ -267,7 +269,11 @@ def _setup_model(self, model: Module) -> Module:
...
+            if _has_meta_device_parameters(model):
+                rank_zero_warn(
+                    "The model is already wrapped in `FSDP` but there are still parameters on the meta device."
+                )
             if "auto_wrap_policy" in self.kwargs:
                 rank_zero_warn(
```
Comment (awaelchli) [inline src/lightning/pytorch/strategies/fsdp.py:278]:
```suggestion
            if "auto_wrap_policy" in self.kwargs:
                # The user has wrapped their submodules manually, don't apply the auto wrap policy.
                rank_zero_warn(
```

**Pull request merged**